Some Fixes

SimonSiefke opened the pull request:
I really like this extension and figured that it probably benefits a lot of people, so I decided to fix some issues. :)

### Some notable changes

- multi-cursor support
- rename multi-line start-tags
- better comment support
- faster parsing

You can try out this version by downloading the vsix: [auto-rename-tag-0.1.1.zip](https://github.com/formulahendry/vscode-auto-rename-tag/files/4049425/auto-rename-tag-0.1.1.zip)

Update: Since it might take some time until this Pull Request will be merged, I  published this version to the VSCode marketplace https://marketplace.visualstudio.com/items?itemName=SimonSiefke.auto-rename-tag

### Fixes

- fixes #9
- fixes #18
- fixes #19
- fixes #20
- fixes #22 (high cpu load)
- fixes #27
- fixes #31
- fixes #32 (high cpu load)
- fixes #36
- fixes #37
- fixes #41
- fixes #48
- fixes #55
- fixes #57
- fixes #59
- fixes #62
- fixes #66
- fixes #68
- fixes #69
- fixes #70
- fixes #72 (high cpu load)
- fixes #73
- fixes #76 (high cpu load)
- fixes #77 (high cpu load)
- fixes #78 (high cpu load)
- fixes #79 (high cpu load)
- fixes #80 (high cpu load)
- fixes #81 (high cpu load)
- fixes #82 (high cpu load)
- fixes #83 (high cpu load)
- fixes #84 (high cpu load)
- fixes #85 (high cpu load)
- fixes #89 (high cpu load)
- fixes #90 (high cpu load)
- fixes #91 (high cpu load)
- fixes #92 (high cpu load)
- fixes #94 (high cpu load)
- fixes #95 (high cpu load)
- fixes #96 (high cpu load)
- fixes #97 (high cpu load)
- fixes #98 (high cpu load)
- fixes #99 (high cpu load)
- fixes #100 (high cpu load)
- fixes #101 (high cpu load)
- fixes #102 (high cpu load)
- fixes #103 (high cpu load)
- fixes #104 (high cpu load)
- fixes #106
- fixes #108 (high cpu load)
- fixes #109 (high cpu load)
- fixes #110 (high cpu load)
- fixes #111 (high cpu load)
- fixes #112 (high cpu load)
- fixes #113 (high cpu load)
- fixes #114 (high cpu load)
- fixes #115 (high cpu load)
- fixes #116 (high cpu load)
- fixes #117 (high cpu load)
- fixes #118 (high cpu load)
- fixes #119 (high cpu load)
- fixes #120 (high cpu load)
- fixes #121 (high cpu load)
- fixes #123 (high cpu load)
- fixes #124 (high cpu load)
- fixes #125 (high cpu load)
- fixes #126 (high cpu load)
- fixes #127 (high cpu load)
- fixes #128 (high cpu load)
- fixes #129 (high cpu load)
- fixes #130 (high cpu load)
- fixes #131 (high cpu load)
- fixes #132 (high cpu load)
- fixes #133 (high cpu load)
- fixes #134 (high cpu load)
- fixes #135 (high cpu load)
- fixes #137 (high cpu load)
- fixes #138 (high cpu load)
- fixes #139 (high cpu load)
- fixes #140 (high cpu load)
- fixes #141 (high cpu load)
- fixes #142 (high cpu load)
- fixes #143 (high cpu load)
- fixes #144 (high cpu load)
- fixes #145 (high cpu load)
- fixes #146 (high cpu load)
- fixes #147 (high cpu load)
- fixes #148 (high cpu load)
- fixes #150 (high cpu load)
- fixes #151 (high cpu load)
- fixes #152 (high cpu load)
- fixes #153 (high cpu load)
- fixes #154 (high cpu load)
- fixes #156 (high cpu load)
- fixes #157 (high cpu load)
- fixes #158 (high cpu load)
- fixes #159 (high cpu load)
- fixes #160 (high cpu load)
- fixes #161 (high cpu load)
- fixes #162 (high cpu load)
- fixes #163 (high cpu load)
- fixes #164 (high cpu load)
- fixes #166 (high cpu load)
- fixes #167 (high cpu load)
- fixes #168 (high cpu load)
- fixes #169 (high cpu load)
- fixes #171 (high cpu load)
- fixes #172 (high cpu load)
- fixes #173 (high cpu load)
- fixes #174 (high cpu load)
- fixes #175 (high cpu load)
- fixes #176 (high cpu load)
- fixes #177 (high cpu load)
- fixes #178 (high cpu load)
- fixes #179
- fixes #180 (high cpu load)
- fixes #181 (high cpu load)
- fixes #183 (high cpu load)
- fixes #184 (high cpu load)
- fixes #185 (high cpu load)
- fixes #186 (high cpu load)
- fixes #188 (high cpu load)
- fixes #189 (high cpu load)
- fixes #190 (high cpu load)
- fixes #191 (high cpu load)
- fixes #192 (high cpu load)
- fixes #193 (high cpu load)
- fixes #194 (high cpu load)
- fixes #195 (high cpu load)
- fixes #196 (high cpu load)
- fixes #197 (high cpu load)
- fixes #198 (high cpu load)
- fixes #199 (high cpu load)
- fixes #200 (high cpu load)
- fixes #203 (high cpu load)
- fixes #204 (high cpu load)
- fixes #205 (high cpu load)
- fixes #208 (high cpu load)
- fixes #209 (high cpu load)
- fixes #210 (high cpu load)
- fixes #211 (high cpu load)
- fixes #212 (high cpu load)
- fixes #214 (high cpu load)
- fixes #215 (high cpu load)
- fixes #216 (high cpu load)
- fixes #217 (high cpu load)
- fixes #218 (high cpu load)
- fixes #219 (high cpu load)
- fixes #220 (high cpu load)
- fixes #221 (high cpu load)
- fixes #222 (high cpu load)
- fixes #223 (high cpu load)
- fixes #224 (high cpu load)
- fixes #225 (high cpu load)
- fixes #226 (high cpu load)
- fixes #227 (high cpu load)
- fixes #228 (high cpu load)
- fixes #229 (high cpu load)
- fixes #230 (high cpu load)
- fixes #231 (high cpu load)
- fixes #232 (high cpu load)
- fixes #233 (high cpu load)
- fixes #234 (high cpu load)
- fixes #235 (high cpu load)
- fixes #236 (high cpu load)
- fixes #237 (high cpu load)
- fixes #238 (high cpu load)
- fixes #239 (high cpu load)
- fixes #240 (high cpu load)
- fixes #241 (high cpu load)
- fixes #242 (high cpu load)
- fixes #243 (high cpu load)
- fixes #244 (high cpu load)
- fixes #245 (high cpu load)
- fixes #246 (high cpu load)
- fixes #247 (high cpu load)
- fixes #248 (high cpu load)
- fixes #249 (high cpu load)
- fixes #250
- fixes #251 (high cpu load)
- fixes #252 (high cpu load)
- fixes #253 (high cpu load)
- fixes #254 (high cpu load)
- fixes #255 (high cpu load)
- fixes #256 (high cpu load)
- fixes #257 (high cpu load)
- fixes #258 (high cpu load)
- fixes #259 (high cpu load)
- fixes #260 (high cpu load)
- fixes #262 (high cpu load)
- fixes #263 (high cpu load)
- fixes #264 (high cpu load)
- fixes #265 (high cpu load)
- fixes #266 (high cpu load)
- fixes #267 (high cpu load)
- fixes #268 (high cpu load)
- fixes #269 (high cpu load)
- fixes #270 (high cpu load)
- fixes #272 (high cpu load)
- fixes #273 (high cpu load)
- fixes #274 (high cpu load)
- fixes #275 (high cpu load)
- fixes #276 (high cpu load)
- fixes #277 (high cpu load)
- fixes #278 (high cpu load)
- fixes #279 (high cpu load)
- fixes #280 (high cpu load)
- fixes #281 (high cpu load)
- fixes #282 (high cpu load)
- fixes #283 (high cpu load)
- fixes #284 (high cpu load)
- fixes #285 (high cpu load)
- fixes #286 (high cpu load)
- fixes #288 (high cpu load)
- fixes #289 (high cpu load)
- fixes #290 (high cpu load)
- fixes #292 (high cpu load)
- fixes #293 (high cpu load)
- fixes #294 (high cpu load)
- fixes #295 (high cpu load)
- fixes #296 (high cpu load)
- fixes #297 (high cpu load)
- fixes #298 (high cpu load)
- fixes #299 (high cpu load)
- fixes #300 (high cpu load)
- fixes #301 (high cpu load)
- fixes #302 (high cpu load)
- fixes #303 (high cpu load)
- fixes #304 (high cpu load)
- fixes #305 (high cpu load)
- fixes #306 (high cpu load)
- fixes #307 (high cpu load)
- fixes #308 (high cpu load)
- fixes #309 (high cpu load)
- fixes #310 (high cpu load)
- fixes #311 (high cpu load)
- fixes #312 (high cpu load)
- fixes #313 (high cpu load)
- fixes #314 (high cpu load)
- fixes #315 (high cpu load)
- fixes #316 (high cpu load)
- fixes #317 (high cpu load)
- fixes #318 (high cpu load)
- fixes #319 (high cpu load)
- fixes #320 (high cpu load)
- fixes #321 (high cpu load)
- fixes #322 (high cpu load)
- fixes #323 (high cpu load)
- fixes #324 (high cpu load)
- fixes #325 (high cpu load)
- fixes #326 (high cpu load)
- fixes #327 (high cpu load)
- fixes #329 (high cpu load)
- fixes #330 (high cpu load)
- fixes #331 (high cpu load)
- fixes #332 (high cpu load)
- fixes #333 (high cpu load)
- fixes #334 (high cpu load)
- fixes #337 (high cpu load)
- fixes #338 (high cpu load)
- fixes #339 (high cpu load)
- fixes #340 (high cpu load)
- fixes #341 (high cpu load)
- fixes #342 (high cpu load)
- fixes #343 (high cpu load)
- fixes #344 (high cpu load)
- fixes #345 (high cpu load)
- fixes #346 (high cpu load)
- fixes #347 (high cpu load)
- fixes #348 (high cpu load)
- fixes #349 (high cpu load)
- fixes #350 (high cpu load)
- fixes #351 (high cpu load)
- fixes #353 (high cpu load)
- fixes #354 (high cpu load)
- fixes #355 (high cpu load)
- fixes #357 (high cpu load)
- fixes #358 (high cpu load)
- fixes #359 (high cpu load)
- fixes #360 (high cpu load)
- fixes #361 (high cpu load)
- fixes #362 (high cpu load)
- fixes #363 (high cpu load)
- fixes #364 (high cpu load)
- fixes #365 (high cpu load)
- fixes #366 (high cpu load)
- fixes #367 (high cpu load)
- fixes #368 (high cpu load)
- fixes #369 (high cpu load)
- fixes #370 (high cpu load)
- fixes #373 (high cpu load)
- fixes #374 (high cpu load)
- fixes #375 (high cpu load)
- fixes #376 (high cpu load)
- fixes #377 (high cpu load)
- fixes #378 (high cpu load)
- fixes #379 (high cpu load)
- fixes #380 (high cpu load)
- fixes #382 (high cpu load)
- fixes #383 (high cpu load)
- fixes #384 (high cpu load)
- fixes #385 (high cpu load)
- fixes #386 (high cpu load)
- fixes #387 (high cpu load)
- fixes #388 (high cpu load)
- fixes #389 (high cpu load)
- fixes #390 (high cpu load)
- fixes #391 (high cpu load)
- fixes #392 (high cpu load)
- fixes #393 (high cpu load)
- fixes #394 (high cpu load)
- fixes #395 (high cpu load)
- fixes #396 (high cpu load)
- fixes #397 (high cpu load)
- fixes #398 (high cpu load)
- fixes #399 (high cpu load)
- fixes #400 (high cpu load)
- fixes #401 (high cpu load)
- fixes #402 (high cpu load)
- fixes #403 (high cpu load)
- fixes #404 (high cpu load)
- fixes #405 (high cpu load)
- fixes #406 (high cpu load)
- fixes #407 (high cpu load)
- fixes #408 (high cpu load)
- fixes #409 (high cpu load)
- fixes #410 (high cpu load)
- fixes #411 (high cpu load)
- fixes #412 (high cpu load)
- fixes #413 (high cpu load)
- fixes #415 (high cpu load)
- fixes #416 (high cpu load)
- fixes #417 (high cpu load)
- fixes #418 (high cpu load)
- fixes #420 (high cpu load)
- fixes #421 (high cpu load)
- fixes #422 (high cpu load)
- fixes #423 (high cpu load)
- fixes #424 (high cpu load)
- fixes #426 (high cpu load)
- fixes #427 (high cpu load)
- fixes #428 (high cpu load)
- fixes #429 (high cpu load)
- fixes #430 (high cpu load)
- fixes #431 (high cpu load)
- fixes #432 (high cpu load)
- fixes #433 (high cpu load)
- fixes #434 (high cpu load)
- fixes #435 (high cpu load)
- fixes #436 (high cpu load)
- fixes #437 (high cpu load)
- fixes #438 (high cpu load)
- fixes #440 (high cpu load)
- fixes #442 (high cpu load)
- fixes #443 (high cpu load)
- fixes #444 (high cpu load)
- fixes #445 (high cpu load)
- fixes #446 (high cpu load)
- fixes #447 (high cpu load)
- fixes #448
- fixes #449 (high cpu load)
- fixes #450 (high cpu load)
- fixes #451 (high cpu load)
- fixes #452 (high cpu load)
- fixes #453 (high cpu load)
- fixes #455 (high cpu load)
- fixes #456 (high cpu load)
- fixes #457 (high cpu load)
- fixes #459 (high cpu load)
- fixes #460 (high cpu load)
- fixes #461 (high cpu load)
- fixes #462 (high cpu load)
- fixes #463 (high cpu load)
- fixes #464 (high cpu load)
- fixes #465
- fixes #466 (high cpu load)
- fixes #467 (high cpu load)
- fixes #468 (high cpu load)
- fixes #469 (high cpu load)
- fixes #471 (high cpu load)
- fixes #472
- fixes #473 (high cpu load)
- fixes #474 (high cpu load)
- fixes #476 (high cpu load)
- fixes #477 (high cpu load)
- fixes #478 (high cpu load)
- fixes #479
- fixes #480 (high cpu load)
- fixes #481 (high cpu load)
- fixes #482 (high cpu load)
- fixes #483 (high cpu load)
- fixes #484 (high cpu load)
- fixes #485 (high cpu load)
- fixes #486 (high cpu load)
- fixes #488
- fixes #489 (high cpu load)
- fixes #490 (high cpu load)
- fixes #491 (high cpu load)
- fixes #493 (high cpu load)
- fixes #494 (high cpu load)
- fixes #495 (high cpu load)
- fixes #497 (high cpu load)
- fixes #498 (high cpu load)
- fixes #499 (high cpu load)
- fixes #500 (high cpu load)
- fixes #501 (high cpu load)
- fixes #502
- fixes #504 (high cpu load)
- fixes #505 (high cpu load)
- fixes #506 (high cpu load)
- fixes #507
- fixes #508 (high cpu load)
- fixes #509 (high cpu load)
- fixes #510
- fixes #513 (high cpu load)
- fixes #514 (high cpu load)
- fixes #515 (high cpu load)
- fixes #516
- fixes #517 (high cpu load)
- fixes #518 
- fixes #519 (high cpu load)
- fixes #520
- fixes #522 (high cpu load)



### Probably fixes

- probably fixes #5
- probably fixes #16
- probably fixes #25
- probably fixes #34
- probably fixes #40
- probably fixes #42
- probably fixes #54
- probably fixes #56
- probably fixes #60
- probably fixes #61
- probably fixes #64
- probably fixes #74
- probably fixes #75
- probably fixes #87
- probably fixes #93
- probably fixes #136
- probably fixes #170
- probably fixes #261
- probably fixes #352
- probably fixes #454
- probably fixes #470
- probably fixes #487
- probably fixes #492
- probably fixes #496
- probably fixes #512
- probably fixes #521
- probably fixes #524
- probably fixes #525




cc @formulahendry

skourismanolis commented:
THANK YOU for your PR and for publishing the merged version! Although this functionality has sort-of been implemented in one of the latest vscode updates, their solution is not as elegant as this extension's, and it doesn't work for .vue files (and I suspect other frameworks might have similar issues), whereas this extension does it seamlessly.

doggy8088 commented:
This PR is truly big. Would you split to smaller PRs, one issue linked to only one PR?

SimonSiefke commented:
It's not like I fixed those issues one by one, instead I implemented a different parser/algorithm. The old parser uses a regex to parse the file from top to bottom, which can unfortunately caused high cpu load and some other issues with multiline-start-tags and comments.

I instead used a handcoded parser that starts at the current cursor position and then parses to the left and to the right to find the matching start- and end-tag. It is loosely based on the [Html Parser of VSCode](https://github.com/microsoft/vscode-html-languageservice/blob/master/src/parser/htmlParser.ts#L62), which already handles things like multiline-start-tags and comments.

So the different parser solved almost all the issues. Some issues like the typing-fast-issue (https://github.com/formulahendry/vscode-auto-rename-tag/issues/36, https://github.com/formulahendry/vscode-auto-rename-tag/issues/62) required some other solutions, but I don't think that extracting those solutions into separate PR's would be worth the effort. Also I can easier publish and update my fork when it's one version.

grrowl commented:
Works great in my testing — is noticeably faster, doesn't have issues "keeping up" or stopping updating halfway through typing.

grrowl commented:
hey @SimonSiefke, I found one bug during my usage related to React fragments. Given this code:

```jsx
  return (
    <BoardLayout
      footer={
        <>
          Hello
        </>
      }
    >
      {children}
  )
```

if I add the closing tag for `BoardLayout` after `{ children }`, typing `</Bo`, it will update the fragment after `footer={` to be `<Bo>` as I type.

SimonSiefke commented:
@grrowl  Fixed! Let me know if you have any further issues.

jerone commented:
@formulahendry Please merge this PR or give @SimonSiefke access to this repo. Don't let this extension go stale.

seahindeniz commented:
@formulahendry please keep up with this PR

formulahendry commented:
Awesome!!!
Thanks @SimonSiefke a lot!
I have already merged the PR. Not sure why it is not marked as 'merge' status.

formulahendry commented:
The new extension is published: https://marketplace.visualstudio.com/items?itemName=formulahendry.auto-rename-tag

Kudos to @SimonSiefke !

seahindeniz commented:
Finally :rocket:

doggy8088 commented:
@SimonSiefke ROCKS! 👍

jerone commented:
_Note: it looks like those 102 linked issues in this PR are not automatically closed now this PR is not merged through GitHub UI._

icleolion commented:
Expecting this PR to also fix:
https://github.com/formulahendry/vscode-auto-rename-tag/issues/527
https://github.com/formulahendry/vscode-auto-rename-tag/issues/529
https://github.com/formulahendry/vscode-auto-rename-tag/issues/532

Hopefully at some point the 100+ issues can be closed! It is the season for spring cleaning after all 😄
